### DISCUSSION TITLE
AIR for Windows - swf-version 38+ falling back to software rendering on DirectX 11

### DISCUSSION THREAD

**Discussion by FliplineStudios** [airsdk/Adobe-Runtime-Support#1754] [Q&A]
In the past I've focused on mobile AIR development, but I'm starting to explore Windows builds with AIR for desktop and ran across an issue, and wondered if anyone had any suggestions: 

I was previously using -swf-version=37 when building apps with FlashDevelop -- which would target the older AIR 26 SDK and its DirectX 9 hardware acceleration on PC (we would do final packaging with 33.1 later on the command line).  We wanted to make use of the ScreenMode additions in AIR 31 though, so I switched to a higher swf-version to have access to those.

Once I changed to any swf-version beyond 37 I noticed much slower performance on desktop, and found that starting with AIR 27 it changed to use DirectX 11 for hardware acceleration instead of DirectX 9 like before.  Looking in Scout, when I build with -swf-version=37 (and renderMode direct or gpu) I'm seeing information about my GPU/Driver/Interface and showing DirectX 9.0, so the older versions are being hardware accelerated.  When I build with -swf-version=38 or higher, I'm not seeing any GPU information in Scout, so I'm assuming it's falling back to software rendering.  This is an older PC I'm testing on, and while dxdiag claims I have DirectX 12 installed and the displays support DirectX 11, it's possible that something with this setup just isn't fully supported and so that's why AIR is falling back to software rendering.  This isn't specifically related to the Harman AIR versions, as I'm seeing the same issue with compiling with AIR 28 with -swf-version 38 or 39 as well.

I was wondering if anyone else has any experience with problems when AIR switched from DirectX 9 to DirectX 11, any thoughts on why it's not being hardware accelerated using DirectX 11, or any suggestions on how to keep this desktop app hardware accelerated?  It sounds like on Windows 7 and below, AIR would still use DirectX 9 on Windows -- I don't suppose there's any setting somewhere to force AIR to always use DirectX 9 regardless of Windows version?  I'm guessing my other option is to stick with swf-version 37 and not use the ScreenMode classes, if I want to be able to support older PCs that may not handle the DirectX 11 rendering correctly.  I have no idea how widespread this issue might be (or if it's just this old PC that has spotty DirectX 11 support), though I'd worry that average end users on older PCs might also have the same problem.

Any thoughts or suggestions would be appreciated!

**Comment by ajwfrost** [comment]
Hi
From the code logic, you're right in terms of the decision when to use Direct3D 11 vs Direct3D 9 (there's also a Direct3D 10 option but I think that would only be hit in some Flash Player variants..). Direct3D 11 is preferred when it's Windows 8 or later, but it switches back to Direct3D 9 if the AIR version is less than AIR version 27 (which is equivalent to SWF version 38).

The thing I can't quite work out, if it's failing in actually setting up the Direct3D 11 device/renderer, is why it doesn't then fall back to Direct3D 9. But this does seem to be a limitation just in terms of how the code is set up, so it may be possible to refactor this and have it try Direct3D 11 and then if that fails, retry with Direct3D 9 before then falling back to software rendering if it's still not getting up.

The technical challenge is that there's quite a lot of initialisation that's already happened here before all the Direct3D stuff kicks off which may need to then be re-done, so we will have a look at how this could work...

**Reply by FliplineStudios** [reply]
Hi Andrew, thanks for the explanation on this!  Ahh okay, I was hoping there was a <forceDirectX9> tag or some loosely-documented feature that I had missed where it could always use DirectX 9.  I'm assuming I can just keep using swf-version 37 to force it into DirectX 9, and still keep packaging with AIR 33.1.x for the same effect?  Everything seems to be working fine with that method so far, I just miss out on any API changes that came after AIR 26...

**Reply by ajwfrost** [reply]
@FliplineStudios we've just added a tag so that you can set your app to use Direct3D 9, as there are other problems we're finding now with Direct3D 11 with some drivers. So hopefully this would then mean that you can build your app with the latest AS3 APIs and not have to limit the SWF version code...
See https://github.com/airsdk/Adobe-Runtime-Support/issues/1963#issuecomment-1206060772
